lib/espresso-runner.js

```diff
@@ -29,40 +29,28 @@ class EspressoRunner {
     this.serverLaunchTimeout = opts.serverLaunchTimeout || ESPRESSO_SERVER_LAUNCH_TIMEOUT;
   }
 
-  async shouldUninstallServer () {
-    if (this.forceEspressoRebuild || !await this.adb.fileExists(TARGET_PACKAGE_CONTAINER)) {
+  async isAppPackageChanged () {
+    if (!await this.adb.fileExists(TARGET_PACKAGE_CONTAINER)) {
+      logger.debug('The previous target application package is unknown');
       return true;
     }
-
-    const tmpRoot = await tempDir.openDir();
-    try {
-      const dstPath = path.resolve(tmpRoot, path.posix.basename(TARGET_PACKAGE_CONTAINER));
-      await this.adb.pull(TARGET_PACKAGE_CONTAINER, dstPath);
-      const previousAppPackage = await fs.readFile(dstPath, 'utf8');
-      logger.debug(`The previous target application package was '${previousAppPackage}'. ` +
-        `The current package is '${this.appPackage}'.`);
-      return previousAppPackage !== this.appPackage;
-    } finally {
-      await fs.rimraf(tmpRoot);
-    }
+    const previousAppPackage = (await this.adb.shell(['cat', TARGET_PACKAGE_CONTAINER])).trim();
+    logger.debug(`The previous target application package was '${previousAppPackage}'. ` +
+      `The current package is '${this.appPackage}'`);
+    return previousAppPackage !== this.appPackage;
   }
 
   async installServer () {
     await this.adb.installOrUpgrade(this.modServerPath, TEST_APK_PKG);
-
-    const tmpRoot = await tempDir.openDir();
-    try {
-      const srcPath = path.resolve(tmpRoot, path.posix.basename(TARGET_PACKAGE_CONTAINER));
-      await fs.writeFile(srcPath, this.appPackage, 'utf8');
-      await this.adb.push(srcPath, TARGET_PACKAGE_CONTAINER);
-      logger.info(`Recorded the target application package '${this.appPackage}' to ${TARGET_PACKAGE_CONTAINER}`);
-    } finally {
-      await fs.rimraf(tmpRoot);
-    }
     logger.info(`Installed Espresso Test Server apk '${this.modServerPath}' (pkg: '${TEST_APK_PKG}')`);
   }
 
   async installTestApk () {
+    if (!this.forceEspressoRebuild && await this.isAppPackageChanged()) {
+      logger.info(`Forcing espresso server rebuild because of changed application package`);
+      this.forceEspressoRebuild = true;
```

Comment:
This seems odd. I don't think the object property should be changed when unnecessary. The fact that rebuild will have to happen is localized to this function, so a local variable would suffice.

Reply:
I think something like the following would be better
```js
let rebuild = this.forceEspressoRebuild;
if (rebuild) {
  logger.debug(`Capability 'forceEspressoRebuild' on.`);
} else if (await this.isAppPackageChanged()) {
  logger.info(`Forcing espresso server rebuild because of changed application package`);
  rebuild = true;
}
if (rebuild && await fs.exists(this.modServerPath)) {
  logger.debug(`Espresso server rebuilt required. Deleting file '${this.modServerPath}'`);
  await fs.unlink(this.modServerPath);
  await this.buildNewModServer();
}
await this.checkAndSignCert(this.modServerPath);
if (rebuild) {
  await this.adb.uninstallApk(TEST_APK_PKG);
  logger.info('Uninstalled the obsolete Espresso server package from the device under test');
}
```

Reply:
makes sense, will change